### DISCUSSION TITLE
Changing patches using "Program Change" midi message

### DISCUSSION THREAD

**Discussion by Stratg** [GeekFunkLabs/fluidpatcher#101] [Q&A]
Hi, 
I use a headless Squishbox and want to use the "Program Change" keys on my midi keyboard controller.
When my keyboard send a "Program Change" message, it doesn't change the patch described in bank file, but change the instrument in the sf2 file. 

When i read headlesspi.py, on the line 44:  

fp.add_router_rule(type='cc', chan=CHAN, par1=SELECT_PATCH, par2=selectspec, patch='select')

I wonder if there is  a way to use the standard 'prog' midi message to change the patch ?

Perhaps there is a way to override using a "bankX.yaml" file ?

Thank you for your help !!

![image](https://github.com/GeekFunkLabs/fluidpatcher/assets/108470745/c2d8e52f-8fd3-4505-ad34-291b58a4a418)

**Comment by albedozero** [comment]
Sure - just change the line you noted to this:

```python
fp.add_router_rule(type='prog', chan=CHAN, par1='0-127*1+1', patch='select')
```

Make sure to also set `SELECT_PATCH=True` so the command above will get processed. The first argument selects program change messages. PC messages only have one parameter, instead of two like note and control change messages, so we use that to select the patch. The `par1` argument just adds 1 to the program change value, since program changes start at 0 but patch numbers start at 1. Values greater than your current number of patches get wrapped around by the code. You'll get a jump in the patch number when your keyboard wraps from 127-0 or 0-127 if your number of patches doesn't divide evenly into 128 - there isn't a neat way around this, but if I think of something I'll post it here. Let me know if this works for you.

**Reply by Stratg** [reply]
Thank you for your help.

I, have made the modification, but it causes crash:

```
pi@squishbox:~ $ sudo systemctl status  squishbox.service
× squishbox.service - Headless Pi Synth
     Loaded: loaded (/etc/systemd/system/squishbox.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Tue 2023-12-12 10:22:44 GMT; 403ms ago
   Duration: 130ms
    Process: 1839 ExecStart=/home/pi/headlesspi.py (code=exited, status=1/FAILURE)
   Main PID: 1839 (code=exited, status=1/FAILURE)
        CPU: 130ms
```

I've tried different syntax, and simplified the script but it the same.  I've tried to write a rule in bank file, but it doesn't work.

However, on Windows Fluidpatcher version, the rule in bank file works fine adding this by example:

```
router_rules:
  - {type: prog, par1:1-127=1-10, patch: select} 
```

It seems the raspberry version doesn't understand the 'prog' instruction.

I will have some other tests.

Thanks a lot!!

**Reply by Stratg** [reply]
Ok, doesn't crash anymore (indentation problem: space vs tabs...) but 
`fp.add_router_rule(type='prog', chan=CHAN, par1='0-127*1+1', patch='select')`
is not working.

The synth capture the "program change" message and make the patch and the program changing.

**Comment by albedozero** [comment]
Aha - the program change message continues to get processed even though the patch has changed. I'm working on #102 to fix this, but for now instead of making the changes to _headlesspi.py_ I suggested above, you would add the following at the beginning of the bank-level `router_rules` in all of the bank files you want to use:

```yaml
router_rules:
- clear
- {type: note}
- {type: cc}
- {type: pbend}
- {type: cpress}
- {type: kpress}
- {type: prog, par1: 0-127*1+1, patch: select}
```

This clears all the default router rules (including program changes), then creates default routing for the other five voice message types, and finally routes program changes to patch selection.

**Comment by Stratg** [comment]
Ok, now it works fine with this modification. It seems better to me to add it in the bank file instead of the python code.
Now I can use an older keyboard with few extra cc keys. This project is really very interesting !
I'll stay tuned for the fix! Thank you
